### ISSUE TITLE
Watcher-Specific Options (namely `deep:true`)

### ISSUE THREAD

**Issue by adamgoose** [yyx990803/vue#974]
Current Version: 0.12.4

Currently, defining a watcher via the Vue constructor does *not* register the watcher as a deep watcher. However, passing `deep: true` to the `watch` property of the constructor options sets all watchers to deep.

It'd be nice to be able to set `deep:true` on a specific watcher, possibly like this:

```js
new Vue({
  watch: {
    'foo': {
      handler: function(val, oldVal) {},
      options: { deep: true }
    }
  }
});
```

This could be accomplished by modifying the `register` function in `/src/instance/events.js` as follows:

```js
function register (vm, action, key, handler) {
  var type = typeof handler
  if (type === 'function') {
    vm[action](key, handler)
  } else if (type === 'object' && action === '$watch') { // Line Added
    vm[action](key, handler.handler, handler.options) // Line Added
  } else if (type === 'string') {
    var methods = vm.$options.methods
    var method = methods && methods[handler]
    if (method) {
      vm[action](key, method)
    } else {
      _.warn(
        'Unknown method: "' + handler + '" when ' +
        'registering callback for ' + action +
        ': "' + key + '".'
      )
    }
  }
}
```

The special case for `action === '$watch'` is necessary, since the `$on` method doesn't support an options argument, thus making this a tad less flexible for any future watchers that may get added. However, it would work!

**Comment by yyx990803** [comment]
Thanks for the suggestion. Landed in 47c52d6bfc59bd17a0f0f1ebc8c77730f9716685 although with some difference:

``` js
watch: {
  'foo': {
    // also support string method names
    handler: 'methodName',
    // no need to nest options here
    deep: true
  }
}
```

**Comment by adamgoose** [comment]
Very cool! Thank you so much. :) 

@taylorzane
